### PR TITLE
feat: Replace Fees DocType with Sales Invoice

### PR DESCRIPTION
**Pull request by RitvikSardana**
### Issue:

Right now Fees DocType is used to for handling Student Fees in the Education Module.

The issues with Fees DocType are:

1. No Payment Terms
2. No Debit Note
3. No Credit Note
4. No support for Statement of Accounts.
and so on....

### Solution:
Creating Sales Invoice instead of Fees solved all of our issues.

### What Changed?

1. User can now create various Fee Schedules from Fee Structure based on Fee Plan. (Demo Below) 

https://github.com/frappe/education/assets/65544983/23603908-2721-4c73-bec5-88984dc2fdb3

2. New child table “Fee Schedule Details” to show the table in the Fee Schedule Modal.

3. Add discounts while creating a fee schedule.
Screenshot 2024-01-24 at 2 51 15 PM

4. Use Link Fields child table to make connections instead of using Dashboard code.

5. Change Status Color in Fee Schedule.

6. Create Sales Invoice Instead of Fees
    a. Custom Fields in Sales Invoice ⇒ Fee Schedule and Student.

7. Create Multiple Sales Invoice via:
    a. Program Enrollment
    b. Fee Schedule

8. Change Program Fee Child Doctype ⇒ Instead of Fee Structure use Fee Schedule (reason: SI records Fee Schedule and not Fee Structure)

9. Create New Customer Group called “Student”.

10. Create New Item Group called “Fee Component”.

11. When a Student Is Created, a customer is also created against the New Student.
    a. Customer Group of the Student Will be “Student”

12. When a Fee Category is created, create an Item Master against the Fee Category.
    a. Item Group of the Item will be “Fee Component”

13. New Education Setting 
    a. “auto_submit_sales_invoice” :  If enabled then Sales Invoice created will be auto submitted.

14. User can't neither create Fees Doctype nor it is searchable.

### CODE REVIEW

File: education/education/doctype/fee_structure/fee_structure.js [hunk 1]
```diff
@@ -43,25 +43,185 @@ frappe.ui.form.on('Fee Structure', {
...
+			title: "Create Fee Schedule",
+			fields:[
+				{
+					label:"Select Fee Plan",
+					fieldname:"fee_plan",
+					fieldtype:"Select",
+					reqd: 1,
+					options:["Monthly","Quarterly","Semi-Annually","Annually","Term-Wise"],
+					change: () => {
+						let fee_plan = dialog.get_value('fee_plan');
```
Comment (pateljannat) [inline education/education/doctype/fee_structure/fee_structure.js]: Put this in a separate function.

File: education/education/doctype/program_enrollment/program_enrollment_dashboard.py [hunk 2]
```diff
@@ -4,9 +4,7 @@
 def get_data():
 	return {
 		"fieldname": "program_enrollment",
-		"transactions": [
-			{"label": _("Course and Fee"), "items": ["Course Enrollment", "Fees"]}
-		],
+		"transactions": [{"label": _("Fees"), "items": ["Fees"]}],
```
Comment (pateljannat) [inline education/education/doctype/program_enrollment/program_enrollment_dashboard.py]: Remove this

File: education/education/doctype/student/student.json [hunk 3]
```diff
@@ -278,11 +287,98 @@
...
+   "fieldtype": "Column Break"
+  },
+  {
+   "fieldname": "default_currency",
+   "fieldtype": "Link",
+   "label": "Billing Currency",
+   "options": "Currency"
+  },
+  {
+   "fieldname": "default_price_list",
```
Comment (pateljannat) [inline education/education/doctype/student/student.json]: Verify this

File: education/education/doctype/student/student_dashboard.py [hunk 4]
```diff
@@ -8,19 +8,6 @@ def get_data():
...
-					"Student Group",
-				],
-			},
-			{"label": _("Assessment"), "items": ["Assessment Result"]},
-			{
-				"label": _("Attendance"),
-				"items": ["Student Attendance", "Student Leave Application"],
-			},
-			{"label": _("Fee"), "items": ["Fees", "Bank Account"]},
+			{"label": _("Fee"), "items": ["Sales Invoice", "Bank Account"]},
```
Comment (pateljannat) [inline education/education/doctype/student/student_dashboard.py:11]: make this a link

**Review by pateljannat — CHANGES_REQUESTED** [review]

**Comment by kodeli-lab** [comment]
Please with the the above implementation how do you track outstanding amount or arrears for the previous semester. Is there a way to automatically include outstanding fees or arrears to the current semester fees. Example, If a student has an outstanding balance for the previous semeters, in the creation of fee schedules for the next semeter, the outstanding fee is automatically added.

**Comment by RitvikSardana** [comment]
both are separate invoices for a student, if the student has not paid the fees, the invoice will show the status as overdue.

**Pull request merged**